File: leetcode/medium/Permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        """
        Generates all permutations of a list without using the itertools module.
        
        :param data: The list of elements to permute.
        :return: A list of lists, where each inner list is a permutation.
        """
        # Base case: If the list has 1 or 0 elements, it is its own only permutation.
        if len(nums) <= 1:
            return [nums]

        # Initialize a list to store all the resulting permutations
        all_perms = []

        # Loop through each element in the list
        for i in range(len(nums)):
            # 1. Pick the current element (the 'head')
            head = nums[i]
            
            # 2. Get the remaining elements (the 'tail')
            # This is everything EXCEPT the element at index i
            tail = nums[:i] + nums[i+1:]
            
            # 3. Recursively find all permutations of the remaining elements
            for perm in permute(tail):
                # 4. Prepend the 'head' element to each permutation of the 'tail'
                # and add the new full permutation to the results list
                all_perms.append([head] + perm)

        return all_perms

File: leetcode/medium/Permutations.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Solution class for generating permutations.
 */
class Solution {

    /**
     * Generates all possible permutations of an array of distinct integers.
     * <p>
     * This implementation uses a backtracking algorithm with in-place swapping
     * to minimize memory overhead and avoid unnecessary array allocations.
     * </p>
     *
     * @param nums The input array of distinct integers.
     * @return A list of lists, where each inner list is a unique permutation.
     */
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        
        // Edge case handling
        if (nums == null || nums.length == 0) {
            return result;
        }

        // Convert int[] to ArrayList<Integer> to facilitate easier swapping using Collections
        List<Integer> numsList = new ArrayList<>(nums.length);
        for (int num : nums) {
            numsList.add(num);
        }

        backtrack(nums.length, numsList, result, 0);
        return result;
    }

    /**
     * Helper method to generate permutations using backtracking.
     *
     * @param n      The length of the input array.
     * @param nums   The current state of the permutation list.
     * @param result The list to store all completed permutations.
     * @param first  The current index we are considering for swapping.
     */
    private void backtrack(int n, List<Integer> nums, List<List<Integer>> result, int first) {
        // Base Case: If the current index 'first' reaches the size of the array,
        // we have a complete permutation.
        if (first == n) {
            // We must create a new ArrayList (deep copy) because 'nums' is passed by reference.
            // If we added 'nums' directly, all entries in 'result' would look identical at the end.
            result.add(new ArrayList<>(nums));
            return;
        }

        for (int i = first; i < n; i++) {
            // Step 1: Place the i-th integer at the 'first' position
            Collections.swap(nums, first, i);

            // Step 2: Recurse to generate permutations for the rest of the array
            backtrack(n, nums, result, first + 1);

            // Step 3: Backtrack (Undo the swap) to restore the original state
            // so the next iteration can try a different number at 'first'.
            Collections.swap(nums, first, i);
        }
    }
}